text/0566-memo-decorator.md

```diff
@@ -0,0 +1,141 @@
+- Start Date: 2019-12-22
+- Relevant Team(s): Ember.js
+- RFC PR: https://github.com/emberjs/rfcs/pull/566
+- Tracking: (leave this empty)
+
+# @memo
+
+## Summary
+
+Add a `@memo` decorator for memoizing the result of a getter based on
+autotracking. In the following example, `fullName` would only recalculate if
+`firstName` or `lastName` is updated.
+
+```js
+class Person {
+  @tracked firstName = 'Jen';
+  @tracked lastName = 'Weber';
+
+  @memo
+  get fullName() {
+    return `${this.firstName} ${this.lastName}`;
+  }
+}
+```
+
+## Motivation
+
+One of the major differences between computed properties and tracked properties
+with autotracking in Octane is that native, autotracked getters do not
+automatically cache their values, where computed properties were cached by
+default. This was an intentional design choice, as the memoization logic for
+computed properties was actually more costly, on average, than rerunning the
+getter in the first place. This was especially true given that computed
+properties would usually only ever be calculated and used once or twice per
+render before being updated.
+
+However, there are absolutely cases where getters _are_ expensive, and their
+values are used repeatedly, so memoization would be very helpful. Strategic,
+opt-in memoization is a useful tool that would help Ember developers optimize
+their apps when relevant, without adding extra overhead unless necessary.
+
+## Detailed design
+
+The `@memo` decorator will be exported from `@glimmer/tracking`, alongside
```

Comment:
I understand that
- this is a very `@glimmer/tracking`-specific feature
- that is deeply connected to the tracking system
- and embedding it there
  - standardizes the implementation across all apps
  - and also significantly reduces the time to market

But just so the question was asked: Was is considered to open up the tracking / tags API instead to allow user-land modules to implement `@memo` and other decorators?

Reply:
Yes! I actually have another RFC in the works to do just this 😄 I wanted to land `@memo` because 1. it's a very small RFC, took half a day to write and I don't think it's super contentious and 2. I do think we should have a conventional solution for this in the framework, enough folks have asked for it that it would be a pain to have to install from the community.

But I 100% agree, we should have public tracking primitives that you could use to build something like `@memo` on top of!